src/com/sun/jna/Structure.java

```diff
@@ -845,23 +855,21 @@ protected void writeField(StructField structField) {
             if (value != null) {
                 // If we've already allocated a native string here, and the
                 // string value is unchanged, leave it alone
-                if (nativeStrings.containsKey(structField.name + ".ptr")
-                    && value.equals(nativeStrings.get(structField.name + ".val"))) {
+                PointerTracking tracking = nativeStrings.get(structField.name);
+                if (tracking != null && value.equals(tracking.value)) {
                     return;
                 }
                 NativeString nativeString = wide
                     ? new NativeString(value.toString(), true)
                     : new NativeString(value.toString(), encoding);
                 // Keep track of allocated C strings to avoid
                 // premature garbage collection of the memory.
-                nativeStrings.put(structField.name, nativeString);
+                nativeStrings.put(structField.name, new PointerTracking(nativeString.getPointer(), nativeString));
                 value = nativeString.getPointer();
             }
             else {
                 nativeStrings.remove(structField.name);
             }
-            nativeStrings.remove(structField.name + ".ptr");
-            nativeStrings.remove(structField.name + ".val");
```

Comment:
The change in general looks very good and the numbers also support that.

The change here also looks sane, but I don't understand why it worked in the first place. Here different keys (once the plain field name (line 857) and once modified field names (863+864).

Do you have an idea what is the reasoning here?

Reply:
I had a look at the history of this class and i think understand better what is going on here.
The map is good for two things:
1. (the initial purpose) keep written strings allocated
2. (added some years later) avoid unchanged strings to be overwritten

Now i can tell i made an error in this PR, the following code will fail with this PR:

´´´

    @Structure.FieldOrder({"value"})
    public static class SimpleStructure extends Structure {
        public String value;
    }
    
    public static void main(String[] args) {
         SimpleStructure sise = new SimpleStructure();
        
         sise.value = "test";       
         sise.write(); // <- create a native string an track it in the map
         sise.read(); // <- overwrites the entry in the map, the native string could be released now
    }
´´´

Even better: there seems to be a bug in the original implementation too, clear calls are not handled at all and lead to unexpected behaviour. e.g. I would expect the following code to print "sise.value => test", it actually prints "sise.value => null" with JNA 5.6.0

´´´

    @Structure.FieldOrder({"value"})
    public static class SimpleStructure extends Structure {
        public String value;
    }

    public static void main(String[] args) {
        SimpleStructure sise = new SimpleStructure();
        
        sise.value = "test";       
        sise.write();
        sise.read();
        
        sise.clear();

        sise.value = "test";       
        sise.write();
        sise.read();
        
        System.out.println("sise.value => " + sise.value);
    }
```
I will update the PR today or tomorrow to handle all cases, including clear calls.